resources/2024-10-dapr-bindings.md

```diff
@@ -0,0 +1,326 @@
+  properties: {
+    application: app.id
+    environment: environment
+    resourceProvisioning: 'manual'
+    type: 'bindings.azure.storagequeues'
+    metadata: {
+      accountName: '<ACCOUNT-NAME>'
+      accountKey: '<ACCOUNT-KEY>'
+      queueName: '<QUEUE-NAME>'
+      direction: 'input'
+    }
+  }
+}
+
+resource demoApp 'Applications.Core/containers@2023-10-01-preview' = {
+  name: 'demo-app'
+  properties: {
+    connections: {
+      queue: {
+        source: jobsQueue.id
+      }
+    }
+  ...
+  }
+}
+```
+
+Independently of the backing service used by the external system, the application can receive and process messages using the created binding.
+
+```go
+
+func handler(w http.ResponseWriter, r *http.Request) {
+  ctx := context.Background()
+  client, _:= dapr.NewClient()
+
+  var messagePayload MyMessageType 
+  _ := json.NewDecoder(r.Body).Decode(&messagePayload)
+  
+  // Process the message
+  // ...
+
+  // Acknowledge the message
+  w.WriteHeader(http.StatusOK)     
+}
+
+func main() {
+	r := mux.NewRouter()
+  // Each time a request is received via the input binding
+  endpoint := fmt.Sprintf("/%s", CONNECTION_QUEUE_COMPONENTNAME)
+	r.HandleFunc(endpoint, handler).Methods("POST", "OPTIONS")
+	http.ListenAndServe(":6002", r)
+}
+```
+
+#### User story 2
+
+As a Radius user, I want my application to send messages to external systems without having direct dependencies on them.
+
+This example uses an external SMTP server to send emails. The SMTP server could be centralized within a company's infrastructure or an external service like SendGrid.
+
+```bicep
+// Output binding to an SMTP server. 
+// https://docs.dapr.io/reference/components-reference/supported-bindings/smtp/
+resource mailing 'Applications.Dapr/bindings@2023-10-01-preview' = {
+  name: 'mailing'
+  properties: {
+    application: app.id
+    environment: environment
+    resourceProvisioning: 'manual'
+    type: 'bindings.smtp'
+    metadata: {...}
+  }
+}
+
+resource demoApp 'Applications.Core/containers@2023-10-01-preview' = {
+  name: 'demoApp'
+  properties: {
+    connections: {
+      mail: {
+        source: mailing.id
+      }
+    }
+  ...
+  }
+}
+```
+
+The application can then send messages to the external system using the created binding.
+
+```go
+ctx := context.Background()
+client, _:= dapr.NewClient()
+// Sending an email with the output binding
+in := &dapr.InvokeBindingRequest{
+    Name:      CONNECTION_MAIL_COMPONENTNAME,
+    Operation: "create",
+    Data: []byte("This is the body of the message"),
+    Metadata: map[string]string{"emailTo": "example@example.net", "subject": "Not fishy"},
+}
+out, _:= client.InvokeBinding(ctx, in)
+```
+
+## Design
+
+### High Level Design
+
+#### Architectural components 
+
+##### Input Binding
+```mermaid
+graph LR
+    Client -->|Request| Engine
+    Engine -->|Deploy| Kubernetes
+    Dapr_Sidecar -->|1 - Resolves| Dapr_Binding
+    Dapr_Sidecar <--> |2 - Long polling / Subscription | Service_X
```

Comment:
(Recap meeting)
A Dapr subscription is indeed another CRD, used to create what's called a declarative subscription.

However, this is not what's reflected in the mermaid diagram. In this case, "subscription" refers to the Dapr sidecar using its integrated SDKs to establish a connection to an external resource. Depending on the resource being listened to, this connection could involve long polling or a subscription to a topic (e.g., subscribing to a RabbitMQ queue using the SDK).

This will be corrected in the diagram, as it is misleading